core/services/workflows/syncer/workflow_registry.go

```diff
@@ -362,6 +355,192 @@ func (w *workflowRegistry) readRegistryEvents(ctx context.Context, don capabilit
 	}
 }
 
+// syncUsingEventStrategy syncs workflow registry contract state by watching for events on the contract.
+// It first loads the workflow metadata from the contract state as WorkflowRegistered events,
+// and then starts a goroutine with one loop for handling the events.
+func (w *workflowRegistry) syncUsingEventStrategy(ctx context.Context, don capabilities.DON, reader ContractReader) {
+	w.lggr.Debugw("Loading initial workflows for DON", "DON", don.ID)
+
+	workflowMetadata, loadWorkflowsHead, err := w.getWorkflowMetadata(ctx, don, reader)
+	// TODO: if this fails, we should still start the event looping
+	if err != nil {
+		w.lggr.Errorw("failed to load initial workflows", "err", err)
+		return
+	}
+
+	w.lggr.Debugw("Rehydrating existing workflows", "len", len(workflowMetadata))
+	for _, workflow := range workflowMetadata {
+		toRegisteredEvent := WorkflowRegistryWorkflowRegisteredV1{
+			WorkflowID:    workflow.WorkflowID,
+			WorkflowOwner: workflow.Owner,
+			DonID:         workflow.DonID,
+			Status:        workflow.Status,
+			WorkflowName:  workflow.WorkflowName,
+			BinaryURL:     workflow.BinaryURL,
+			ConfigURL:     workflow.ConfigURL,
+			SecretsURL:    workflow.SecretsURL,
+		}
+		err := w.handler.Handle(ctx, workflowAsEvent{
+			Data:      toRegisteredEvent,
+			EventType: WorkflowRegisteredEvent,
+		})
+		if err != nil {
+			w.lggr.Errorf("failed to handle workflow registration: %s", err)
+		}
+	}
+
+	// Poll for all workflow related events
+	ets := []WorkflowRegistryEventType{
+		ForceUpdateSecretsEvent,
+		WorkflowActivatedEvent,
+		WorkflowDeletedEvent,
+		WorkflowPausedEvent,
+		WorkflowRegisteredEvent,
+		WorkflowUpdatedEvent,
+	}
+
+	w.readRegistryEvents(ctx, ets, don, reader, loadWorkflowsHead.Height)
+}
+
+// handleWorkflowMetadataAsEvents compares workflow metadata state against the engine registry's state.
+// Differences are handled by the event handler by creating events.
+func (w *workflowRegistry) handleWorkflowMetadataAsEvents(ctx context.Context, workflowMetadata []GetWorkflowMetadata, donID uint32) {
+	// Keep track of which of the engines in the engineRegistry have been touched
+	engineKeyMap := map[string]bool{}
+
+	for _, wfMeta := range workflowMetadata {
+		engine, getErr := w.engineRegistry.Get(EngineRegistryKey{Owner: wfMeta.Owner, Name: wfMeta.WorkflowName})
```

Comment:
One very important implication of this is that this is the very last thing that we should do in the registration handler; ie. no other errors can come after it